src/basic_operations.jl

```diff
@@ -4,30 +4,31 @@
 
 # Unary
 @inline Base.:+{T <: AbstractTensor}(S::T) = S
-@inline Base.:-{T <: AbstractTensor}(S::T) = map(-, S)
+@inline Base.:-{T <: AbstractTensor}(S::T) = _map(-, S)
 
 # Binary
-@inline Base.:+{order, dim, T}(S1::Tensor{order, dim, T}, S2::Tensor{order, dim, T}) = map(+, S1, S2)
-@inline Base.:+{order, dim, T}(S1::SymmetricTensor{order, dim, T}, S2::SymmetricTensor{order, dim, T}) = map(+, S1, S2)
-@inline Base.:+{order, dim}(S1::AbstractTensor{order, dim}, S2::AbstractTensor{order, dim}) = +(promote(S1, S2)...)
-@inline Base.:-{order, dim, T}(S1::Tensor{order, dim, T}, S2::Tensor{order, dim, T}) = map(-, S1, S2)
-@inline Base.:-{order, dim, T}(S1::SymmetricTensor{order, dim, T}, S2::SymmetricTensor{order, dim, T}) = map(-, S1, S2)
-@inline Base.:-{order, dim}(S1::AbstractTensor{order, dim}, S2::AbstractTensor{order, dim}) = -(promote(S1, S2)...)
+@inline Base.:+{order, dim}(S1::Tensor{order, dim}, S2::Tensor{order, dim}) = _map(+, S1, S2)
+@inline Base.:+{order, dim}(S1::SymmetricTensor{order, dim}, S2::SymmetricTensor{order, dim}) = _map(+, S1, S2)
+@inline Base.:-{order, dim}(S1::Tensor{order, dim}, S2::Tensor{order, dim}) = _map(-, S1, S2)
+@inline Base.:-{order, dim}(S1::SymmetricTensor{order, dim}, S2::SymmetricTensor{order, dim}) = _map(-, S1, S2)
+@inline Base.:-{order, dim}(S1::AnyTensor{order, dim}, S2::AnyTensor{order, dim}) =  ((SS1, SS2) = promote(S1, S2); _map(-, SS1, SS2))
+@inline Base.:+{order, dim}(S1::AnyTensor{order, dim}, S2::AnyTensor{order, dim}) =  ((SS1, SS2) = promote(S1, S2); _map(+, SS1, SS2))
 
-@inline Base.:+(S1::AbstractTensor, S2::AbstractTensor) = throw(DimensionMismatch("dimension and order must match"))
-@inline Base.:-(S1::AbstractTensor, S2::AbstractTensor) = throw(DimensionMismatch("dimension and order must match"))
+@inline Base.:*(S::AbstractTensor, n::Number) = _map(x -> x*n, S)
+@inline Base.:*(n::Number, S::AbstractTensor) = _map(x -> n*x, S)
+@inline Base.:/(S::AbstractTensor, n::Number) = _map(x -> x/n, S)
 
-@inline Base.:*(S::AbstractTensor, n::Number) = map(x->(x*n), S)
-@inline Base.:*(n::Number, S::AbstractTensor) = map(x->(n*x), S)
-@inline Base.:/(S::AbstractTensor, n::Number) = map(x->(x/n), S)
+Base.:+(S1::AbstractTensor, S2::AbstractTensor) = throw(DimensionMismatch("dimension and order must match"))
+Base.:-(S1::AbstractTensor, S2::AbstractTensor) = throw(DimensionMismatch("dimension and order must match"))
 
 # map implementations
-@inline function Base.map{T <: AbstractTensor}(f, S::T)
+@inline function _map{T <: AbstractTensor}(f, S::T)
     return apply_all(S, @inline function(i) @inboundsret f(S.data[i]); end)
 end
 
-@inline Base.map{order, dim, T1, T2}(f, S1::AbstractTensor{order, dim, T1}, S2::AbstractTensor{order, dim, T2}) = ((SS1, SS2) = promote(S1, S2); map(f, SS1, SS2))
+_map{order, dim}(f, S1::Tensor{order, dim}, S2::SymmetricTensor{order, dim}) = ((SS1, SS2) = promote(S1, S2); _map(f, SS1, SS2))
```

Comment:
This and the next line is not used anymore (see test-coverage) since they are replaced by 14-15 with `AnyTensor`.
Related: Should we have promotion methods that promote `SymmetricTensor` -> `Tensor` but does not promote eltype?